designs/atomic-context-api/atomic-context-api.md

```diff
@@ -0,0 +1,372 @@
+```javascript
+incr(scope, key, amount, (error, value) => {
+    // If an error occurred in process, pass the error object as first argument.
+    // If the process was succeed, pass null as first argument and the incremented value as second argument.
+});
+decr(scope, key, amount, (error, value) => {
+    // If an error occurred in process, pass the error object as first argument.
+    // If the process was succeed, pass null as first argument and the decremented value as second argument.
+});
+```
+
+if the value of the key or `amount` argument is not a number, throw an error.
+
+##### Uasge in a Function node
+<details>
+
+```javascript
+global.incr("count");
+// Increase the value of `count` by 1 in `default` store.
+// And return the value after increment.
+
+global.incr("count", 5);
+// Increase the value of `count` by 5 in `default` store.
+// And return the value after increment.
+
+global.incr("count", "file");
+// Increase the value of `count` by 1 in `file` store.
+// And return the value after increment.
+
+global.incr("count", 5, "file");
+// Increase the value of `count` by 5 in `file` store.
+// And return the value after increment.
+
+global.incr("count", (err, value) => {
+// Increase the value of `count` by 1 in `default` store.
+// Then invoke callback and pass an error object or the value after increment.
+});
+
+global.incr("count", 5, (err, value) => {
+// Increase the value of `count` by 5 in `default` store.
+// Then invoke callback and pass an error object or the value after increment.
+});
+
+global.incr("count", "file", (err, value) => {
+// Increase the value of `count` by 1 in `file` store.
+// Then invoke callback and pass an error object or the value after increment.
+});
+
+global.incr("count", 5, "file", (err, value) => {
+// Increase the value of `count` by 5 in `file` store.
+// Then invoke callback and pass an error object or the value after increment.
+});
+```
+
+</details>
+
+##### Uasge in Other Nodes
+
+TBD
+
+It would be useful that some nodes (e.g. Change Node) can call atomic api.    
+Currently change node can incr/decr with JSONata but it is not atomic.  
+ 
+For example:
+  ```json
+  [{"id":"cd1ed2c0.4f3e7","type":"change","z":"92e079a4.8d9f78","name":"Increment global.count","rules":[{"t":"set","p":"count","pt":"global","to":"$globalContext(\"count\") ? $globalContext(\"count\") + 1 : 1","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":420,"y":60,"wires":[[]]},{"id":"af7472b9.69e3f","type":"inject","z":"92e079a4.8d9f78","name":"","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":220,"y":60,"wires":[["cd1ed2c0.4f3e7"]]}]
+  ```
+
+##### Discussion
+
+1. Is `decr` needed ?  
+   Some databases(e.g. MongoDB, Couchbase see [here](#Atomic%20API%20provided%20by%20NoSQL%20databases)) do not provide `decr` but can decrease the value using `incr` with negative number. Is `decr` needed  for Context API ? 
+
```

Comment:
I think that the context API of function node needs `incr` and` decr` for users to easily see what the code does,
but ,for the Context Store API, only `incr` would be enough.

I assume that the codes of `incr` and `decr` will be almost the same, and the difference is whether to use increment or decrement operator.
Therefore, implementing `incr` and` decr` is redundant and makes it cumbersome to manage code.

Reply:
Agreed - lets stick with just `incr`. If there's a consensus on that, please remove the Discussion section from the doc as the question has been answered. We can then get this merged as an agreed design.

The 'next steps' you've identified can be added as later pull-requests.